src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -1442,16 +1442,45 @@ public Solution WithAnalyzerConfigDocumentTextLoader(DocumentId documentId, Text
+        /// assuming a background compiler is busy building this compilations.
+        /// <para/> A compilation for the project containing the specified document id will be guaranteed to exist with
+        /// at least the syntax tree for the document.
+        /// <para/> This not intended to be the public API, use Document.WithFrozenPartialSemantics() instead.
         /// </summary>
         internal Solution WithFrozenPartialCompilationIncludingSpecificDocument(DocumentId documentId, CancellationToken cancellationToken)
         {
-            var newCompilationState = _compilationState.WithFrozenPartialCompilationIncludingSpecificDocument(documentId, cancellationToken);
+            // in progress solutions are disabled for some testing
+            if (this.Services.GetService<IWorkspacePartialSolutionsTestHook>()?.IsPartialSolutionDisabled == true)
+                return this;
+
+            var currentCompilationState = _compilationState;
+            var currentDocumentState = this.SolutionState.GetRequiredDocumentState(documentId);
+
+            // We want all linked versions of this document to also be present in the frozen solution snapshot (that way
+            // features like 'completion' can see that there are linked docs and give messages about symbols not being
+            // available in certain project contexts). We do this in a slightly hacky way for perf though. Specifically,
+            // instead of parsing *all* the sibling files (which can be expensive, especially for a file linked in many
+            // projects/tfms), we only parse this single tree.  We then use that same tree across all siblings.  That's
+            // technically inaccurate, but we can accept that as the primary purpose of 'frozen partial' is to get a
+            // snapshot *fast* that is allowed to be *inaccurate*.
+            //
+            // We currently do not have any features that both use frozen-partial semantics *and* which would deeply
+            // care about examining sibling files and having them be parsed perfectly.
+            //
+            // Note: this is very different from the logic we have in the workspace to 'UnifyLinkedDocumentContents'. In
+            // that case, we only share trees when completely safe and accurate to do so (for example, where no
+            // directives are involved).  As that is used for the real solution snapshot, it must be correct.  The
+            // frozen-partial snapshot is different as it is a fork that is already allowed to be inaccurate for perf
+            // reasons (for example, missing trees, or missing references).
+            //
+            // The 'force' flag here allows us to share the doc contents even in the case where correctness might be
+            // violated.
+            var allDocumentIds = this.SolutionState.GetRelatedDocumentIds(documentId);
```

Comment:
I think GetRelatedDocumentIds will include documentId in the returned set, so you might want to exclude that or else we're calling WithDocumentContentsFrom() on itself. Looking at the implementation I think it does handle that already and won't really do anything, but might be best to be explicit.

Reply:
i'll comment on this.  but i'm def ok with including the doc itself, and having it bail quickly.

Reply:
doc'ed